packages/@expo/cli/src/localModules/generation.ts

```diff
@@ -0,0 +1,362 @@
+    path.resolve(projectRoot, 'modules'),
+    path.resolve(projectRoot, 'modules', './**'),
+    path.resolve(projectRoot, 'modules', './**/*'),
+  ];
+}
+
+export async function startModuleGenerationAsync({
+  projectRoot,
+  metro,
+}: ModuleGenerationArguments) {
+  const dotExpoDir = ensureDotExpoProjectDirectoryInitialized(projectRoot);
+  const { exp } = getConfig(projectRoot);
+  const filesWatched = new Set<string>();
+
+  const fileExcluded = (absolutePath: string) => {
+    for (const glob of excludePathsGlobs(projectRoot)) {
+      if (path.matchesGlob(absolutePath, glob)) {
+        return true;
+      }
+    }
+    return false;
+  };
+
+  createFreshMirrorDirectories(projectRoot);
+
+  process.env.EXPO_ROUTER_APP_ROOT = path.join(
+    projectRoot,
+    getRouterDirectoryModuleIdWithManifest(projectRoot, exp)
+  );
+
+  const removeFileAndEmptyDirectories = (absoluteFilePath: string) => {
+    if (fs.lstatSync(absoluteFilePath).isSymbolicLink()) {
+      fs.unlinkSync(absoluteFilePath);
+    } else {
+      fs.rmSync(absoluteFilePath);
+    }
+    let dirNow: string = path.dirname(absoluteFilePath);
+    while (fs.readdirSync(dirNow).length === 0 && dirNow !== dotExpoDir) {
+      fs.rmdirSync(dirNow);
+      dirNow = path.dirname(dirNow);
+    }
+  };
+
+  const onRemoveAppFile = (absoluteFilePath: string) => {
+    const { moduleTypesFilePath, moduleExportPath, viewExportPath, viewTypesFilePath } =
+      typesAndLocalModulePathsForFile(projectRoot, absoluteFilePath);
+
+    filesWatched.delete(absoluteFilePath);
+    if (!fileWatchedWithAnyNativeExtension(absoluteFilePath, filesWatched)) {
+      removeFileAndEmptyDirectories(moduleTypesFilePath);
+      removeFileAndEmptyDirectories(moduleExportPath);
+      removeFileAndEmptyDirectories(viewExportPath);
+      removeFileAndEmptyDirectories(viewTypesFilePath);
+    }
+  };
+
+  const metroWatchKotlinAndSwiftFiles = async ({
+    projectRoot,
+    metro,
+    eventTypes = ['add', 'change', 'delete'],
+  }: {
+    metro: MetroServer | null;
+    projectRoot: string;
+    eventTypes?: string[];
+  }) => {
+    const watcher = metro?.getBundler().getBundler().getWatcher();
+
+    const listener = async ({
+      eventsQueue,
+    }: {
+      eventsQueue: {
+        filePath: string;
+        metadata?: {
+          type: 'f' | 'd' | 'l'; // Regular file / Directory / Symlink
+        } | null;
+        type: string;
+      }[];
+    }) => {
+      for (const event of eventsQueue) {
+        if (
+          eventTypes.includes(event.type) &&
+          event.metadata?.type !== 'd' &&
+          !/node_modules/.test(event.filePath) &&
+          /\.(kt|swift)$/.test(event.filePath) &&
+          !fileExcluded(event.filePath) &&
+          fileSupposedToBeWatched(projectRoot, event.filePath)
+        ) {
+          const { filePath } = event;
+          if (event.type === 'add') {
+            addNewFile(projectRoot, filePath, filesWatched);
```

Comment:
why is one `addNewFile` and the other `onRemoveAppFile` – either both are event handlers or both are actions, no?